fairscale/experimental/nn/sync_batchnorm.py

```diff
@@ -138,22 +118,27 @@ def __init__(
     ) -> None:
         super().__init__(*args, **kwargs)  # type: ignore
         self._process_group = process_group if process_group is not None else dist.group.WORLD
+        self.saved_for_2nd_fwd: Optional[Tuple] = None
 
-    def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
+    def forward(self, input: Tensor) -> Tensor:  # type: ignore
         if not dist.is_initialized() or not self.training:
             return super().forward(input)
 
+        if not hasattr(self, "_checkpoint_fwd_counter") or self._checkpoint_fwd_counter == 0:
```

Comment:
The `_checkpoint_fwd_counter` currently is used by FSDP for the case where a module is forwarded multiple times in a single outer forward pass. In SyncBN's case, do we actually only want to keep the stats once in that case?

i.e.:

```
class Layer():
    self.bn = sync_bn()
    def forward():
       ...
       x = self.bn(x)
       ...

class Mod():
       self.layer = Layer()
       def forward():
             return layer(layer(x))
```

In the above case, the bn inside layer is called 4 times if `Mod` is wrapped by a checkpoint. In torch BN's case, the running stats will be kept in 2 of the 4 times. In this change, using `_checkpoint_fwd_counter`, however, it will be kept in only 1 out of 4 times. I think 2 is probably the right thing to do.

Maybe, this BN should keep a `track_running_stats` flag like torch's BN? That way, checkpoint_wrapper will just use `patch_batchnorm` to toggle that flag to get the right behavior?

Reply:
This BN does support track_running_stats. I verified parity is correct for a single forward pass. Did not think about 2 forward passes. Will add a test and code for that.